src/lara-app/components/runtime.tsx

```diff
@@ -79,34 +84,50 @@ export const RuntimeComponent = ({experiment, runKey, firebaseJWT, setError, def
 
   // get experiment data when loaded
   useEffect(() => {
-    firebase
-      .firestore()
-      .collection(`runs/${runKey}/experiments`)
-      .orderBy("createdAt", "desc")
-      .limit(1)
-      .onSnapshot(snapshot => {
-        setQueriedFirestore(true);
-        experimentRef.current = snapshot.docs[0]?.data?.().experiment;
-        const newData = experimentRef.current?.data as IExperimentData | undefined;
-        setExperimentData(newData);
-
-        // if there is no data force an upload
-        setDisplayQrAndMaybeRegenerateQR((newData === undefined) && !reportMode);
-      }, (err) => {
-        setError(err);
-      });
+    if (runKey) {
+      firebase
+        .firestore()
+        .collection(`runs/${runKey}/experiments`)
+        .orderBy("createdAt", "desc")
+        .limit(1)
+        .onSnapshot(snapshot => {
+          setQueriedFirestore(true);
+          experimentRef.current = snapshot.docs[0]?.data?.().experiment;
+          const newData = experimentRef.current?.data as IExperimentData | undefined;
+          setExperimentData(newData);
+
+          // if there is no data force an upload
+          setDisplayQrAndMaybeRegenerateQR((newData === undefined) && !reportOrPreviewMode);
+        }, (err) => {
+          setError(err);
+        });
+    }
   }, []);
 
   // re-generate QR code if showing
   useEffect(() => {
     const interval = setInterval(() => {
-      if (displayingCode.current && !reportMode) {
-        generateQRCode();
+      if (displayingCode.current && !reportOrPreviewMode && runKey && firebaseJWT) {
+        generateQRCode({runKey, firebaseJWT});
       }
     }, UPDATE_QR_INTERVAL);
     return () => clearInterval(interval);
   }, []);
 
+  useEffect(() => {
+    if (containerRef.current) {
+      const checkHeight = () => {
+        const {height} = containerRef.current ? containerRef.current.getBoundingClientRect() : { height: 0 };
+        if (height > 0) {
+          setHeight(height);
+        }
+      };
+
+      const ro = new ResizeObserver(checkHeight);
+      ro.observe(containerRef.current);
+    }
+  }, [containerRef.current]);
+
```

Comment:
This looks like a good candidate for a reusable `useHeightObserver()` hook.

Reply:
I was thinking the same thing as I wrote it.